[FIX] account_invoice_inter_company: missing migration script

chienandalu opened the pull request:
- Without the migration script we lose the link between invoices, so if
we modified the origin we could be duplicating the destination invoice!

- Now when we've posted a move, we can set it back to draft to edit some
things and the repost it. In that case, we don't want to recreate the
destination invoice if it's already posted and maybe even payed.

- As invoices and the classic move entries are the same model now, we
should filter the later ones to avoid triggering multi-company entries, which
aren't supported.

cc @Tecnativa TT30195

ping @pedrobaeza @carlosdauden

chienandalu committed:
[FIX] account_invoice_inter_company: missing migration script

Without the migration script we loose the link between invoices, so if
we modified the origin we could be duplicating the destination invoice!

TT30195

chienandalu commented:
The runbot warning seems to be on `product_multi_company`

pedrobaeza commented on account_invoice_inter_company/migrations/13.0.3.0.0/post-migration.py:
```diff
@@ -0,0 +1,23 @@
+            FROM account_invoice ai
+                JOIN account_invoice ai2 ON ai.auto_invoice_id = ai2.id
+                JOIN account_move am2 ON am2.old_invoice_id = ai2.id
+            WHERE am.old_invoice_id = ai.id AND ai.auto_generated IS NOT NULL
```
```suggestion
            WHERE am.old_invoice_id = ai.id AND ai.auto_generated IS True
```
or 
```suggestion
            WHERE am.old_invoice_id = ai.id AND ai.auto_invoice_id IS NOT NULL
```

chienandalu replied:
Ok, changes done

chienandalu committed:
[FIX] account_invoice_inter_company: draft button

Now we can set an invoice to draft to perform modifications. We want to
avoid it though when the destination move is already posted as we could
end with substantial differences between them.

We want to avoid modifiying the generated invoice if the total amount
differs.

TT30191

chienandalu commented:
Added a commit with this fix:

Now when we've posted a move, we can set it back to draft to edit some
things and the repost it. In that case, we don't want to recreate the
destination invoice if it's already posted and maybe even payed.

pedrobaeza commented:
But I think we should forbid such edition, as there's the possibility of having different amounts...

chienandalu commented:
> But I think we should forbid such edition, as there's the possibility of having different amounts...

Yes, but maybe the change is a trivial one. Maybe we should warn the user about the destination invoice an let them decide?

pedrobaeza commented:
No, for me it's clear: if the destination invoice is posted, then you can unpost the source one. If not posted, then you can do it, and then regenerate the other side one.

pedrobaeza commented on account_invoice_inter_company/models/account_move.py:
```diff
@@ -197,6 +197,24 @@ def _prepare_invoice_data(self, dest_company):
 
+    def button_draft(self):
+        for move in self:
+            company = move._find_company_from_invoice_partner()
```
Why the need of searching for the company? Isn't `move.auto_generated` enough to know that a company is linked?

chienandalu replied:
Well, indeed that `auto_generated` field is not very useful, as we can deduce that from the `auto_invoice_id` already.

chienandalu commented:
> No, for me it's clear: if the destination invoice is posted, then you can unpost the source one. If not posted, then you can do it, and then regenerate the other side one.

Review the new changes. Anyway, I wonder what chance could we give to the user to make legitimate data change that requieres to set draft the origin invoice.

Also, there's never a check in the opposite direction, so we could be unbalancing the changes from the other side.

pedrobaeza commented on account_invoice_inter_company/models/account_move.py:
```diff
@@ -197,6 +197,24 @@ def _prepare_invoice_data(self, dest_company):
+                            "invoice that's in posted state.\n"
+                            "Invoice %s to %s"
+                        )
+                        % (company.name, inter_invoice_posted.name)
```
The order of the parameters is switched and you don't need `company`, just `inter_invoice_posted.partner_id.display_name`.

chienandalu replied:
sure :+1:

pedrobaeza commented:
Yes, there's no good option here indeed. I think you can prevent any manual modification on an auto invoice.

chienandalu commented:
> Yes, there's no good option here indeed. I think you can prevent any manual modification on an auto invoice.

Should we have a field whitelist at least?

pedrobaeza commented:
I think we can do the contrary: only forbid if a different amount is detected.

chienandalu commented:
Ok, let's see now (I should add some tests)

pedrobaeza commented on account_invoice_inter_company/models/account_move.py:
```diff
@@ -214,6 +233,29 @@ def button_cancel(self):
+    def write(self, vals):
+        res = super().write(vals)
+        for move in self.filtered(
+            lambda x: x.sudo().auto_invoice_id and x.state == "posted"
```
Remove the posted check. In fact, you can only modify an invoice in draft state.

chienandalu commented:
Please review again

chienandalu committed:
[FIX] account_invoice_intercompany: filter moves with type entry

As invoices are merged into the account_move table, we have to filter
those which aren't invoices (entries and receipts)

TT30338

pedrobaeza commented on account_invoice_inter_company/models/account_move.py:
```diff
@@ -33,7 +33,8 @@ def _find_company_from_invoice_partner(self):
         res = super().action_post()
-        for src_invoice in self:
+        # Intercompany account entries aren't supported
+        for src_invoice in self.filtered(lambda x: x.type != "entry"):
```
This should be explicit to the 4 types, as there can be also other types:

https://github.com/odoo/odoo/blob/13.0/addons/account/models/account_move.py#L112

chienandalu replied:
Changes done

pedrobaeza commented:
@carlosdauden please review

pedrobaeza commented:
/ocabot merge patch

pedrobaeza commented:
Manually cherry-picked last 2 commits for v14 in 3bc08cbb6ba93124c2e16eb37f35ecc297954058 and 3b7264711f036f90d7a0c6d4b01f7472b8d6cfb2, and also added the missing test patch: 5686a4f9fee10ea340f250222af031df17ff36b3
